src/doc/trpl/dining-philosophers.md

```diff
@@ -0,0 +1,942 @@
+}
+
+struct Table {
+    forks: Vec<Mutex<bool>>,
+}
+
+fn main() {
+    let (done_tx, done_rx) = mpsc::channel();
+
+    let table = Arc::new(Table { forks: vec![
+        Mutex::new(true),
+        Mutex::new(true),
+        Mutex::new(true),
+        Mutex::new(true),
+        Mutex::new(true),
+    ]});
+
+    let philosophers = vec![
+        Philosopher::new("Baruch Spinoza", done_tx.clone(), 0, 1),
+        Philosopher::new("Gilles Deleuze", done_tx.clone(), 1, 2),
+        Philosopher::new("Karl Marx", done_tx.clone(), 2, 3),
+        Philosopher::new("Friedrich Nietzsche", done_tx.clone(), 3, 4),
+        Philosopher::new("Michel Foucault", done_tx.clone(), 0, 4),
+    ];
+
+    let handles: Vec<_> = philosophers.into_iter().map(|p| {
+        let table = table.clone();
+
+        thread::spawn(move || {
+            p.eat(&table);
+        })
+    }).collect();
+
+    for _ in 0..5 {
+        done_rx.recv().unwrap();
+    }
+
+    for h in handles {
+        h.join().ok().expect("Couldn't join a thread.");
+    }
+}
+```
+
+Lots of changes! However, with this iteration, we’ve got a working program.
+Let’s go over the details:
+
+```rust
+use std::sync::{Mutex, Arc};
+```
+
+We’re going to use another structure from the `std::sync` package: `Arc<T>`.
+We’ll talk more about it when we use it.
+
+```rust
+struct Philosopher {
+    name: String,
+    done: Sender<bool>,
+    left: usize,
+    right: usize,
+}
+```
+
+We need to add two more fields to our `Philosopher`. Each philosopher is going
+to have two chopsticks: the one on their left, and the one on their right.
+We’ll use the `usize` type to indicate them, as it’s the type that you index
+vectors with. These two values will be the indexes into the `forks` our `Table`
+has.
+
+```rust
+fn new(name: &str, done: Sender<bool>, left: usize, right: usize) -> Philosopher {
+    Philosopher {
+        name: name.to_string(),
+        done: done,
+        left: left,
+        right: right,
+    }
+}
+```
+
+We now need to construct those `left` and `right` values, so we add them to
+`new()`.
+
+```rust
+fn eat(&self, table: &Table) {
+    let _left = table.forks[self.left].lock()
+        .ok().expect("Couldn't aquire left mutex");
+    let _right = table.forks[self.right].lock()
+        .ok().expect("Couldn't aquire right mutex");
+
+    println!("{} is eating.", self.name);
+
+    thread::sleep_ms(1000);
+
+    self.done();
+}
+```
+
+We have two new logical lines split over four text lines. We’ve also added an
+argument, `table`. We access the `Table`’s list of forks, and then use
+`self.left` and `self.right` to access the fork at that particular index. That
+gives us access to the `Mutex` at that index, and we call `lock()` on it.
+If the mutex is currently being accessed by someone else, we’ll block until
+it becomes available.
+
+The call to `lock()` might fail, and if it does, we want to crash with a good
+error message, so we use the trusty `ok().expect()` pattern from before.
```

Comment:
Can this also elaborate on what the error conditions are here? The specific error case here is that a lock is "poisoned" if a thread panics while the lock is held, and these `ok().expect()` calls are asserting that the lock has not been poisoned. I don't think you'd want to go into too much detail here about poisoning, but it would be nice to link to the docs explaining it.

I'm also somewhat worried about switching back and forth between `unwrap()` and `ok().expect()`